exploration/0475-overriding-extending-namespacing.md

```diff
@@ -0,0 +1,240 @@
+options to use in a given development environment.
+Each function, option, or option value extension needs to work as seamlessly as possible
+with other add-ons and with the built-in functionality.
+
+To that end, we need to define how externally-authored functions appear in a _message_;
+how externally-authored function options (and their values) can be supported;
+and what, if any, effects this has on the namespace of functions and options.
+
+## Background
+
+_What context is helpful to understand this proposal?_
+
+One example of potential add-on functionality that can help readers understand this proposal
+is the use of _skeletons_ for date and number formatting.
+
+The JavaScript `Intl.DateTimeFormat` API provides for the customization of date and time values
+through the use of "option bags". For example:
+
+```js
+new Intl.DateTimeFormat("en-GB", {
+  weekday: "long",
+  year: "numeric",
+  month: "short",
+  day: "numeric",
+  hour: "numeric",
+  minute: "numeric",
+});
+```
+
+This might be expressed in MessageFormat as:
+
+```
+Today is {$today :datetime weekday=long year=numeric month=short day=numeric hour=numeric minute=numeric}
+```
+
+The ICU family of libraries provide a shorthand mechanism called _skeletons_ for accessing
+date and time format options without needing a verbose list of options.
+The same message might look like this with a skeleton:
+
+```
+Today is {$today :datetime skeleton=EEEEyMdjm}
+```
+
+Skeletons are not proposed for inclusion in the default registry
+because they are not universally available in all datetime formatting
+libraries.
+
+## Use-Cases
+
+_What use-cases do we see? Ideally, quote concrete examples._
+
+- Developers need to add options to the base functions to suit local needs.
+  For example, ICU's skeletons as part of the `:datetime` function
+  Support for this option needs to be specified for local implemented versions.
+
+- Developers want to write a function and access it from messages.
+
+- Developers want to import 3rd party formatting packages and use the package's
+  features from within messages.
+
+- Developers wish to import two or more formatting packages
+  and these might have the same-named functions.
+  For example, there might be both an HTML `p` and TTS `p`
+  function.
+
+- Translators and tools would like a machine-readable way to find out the names
+  and option values for add-on packages.
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+- Developers must be able to write functions that do not later collide with items in the default registry.
+- Developers must be able to write function add-ons that do not later collide with items in the default registry.
+- Users should be able to tell visually when an add-on feature has been used vs. a built-in
+- Users should be able to resolve conflicts between add-on packages that use the same
```

Comment:
All of these requirements combined mean that developer are in control of choosing the namespace name. I think that's by design and desirable, but we should address in the doc the negative impact this can have on portability of messages.